File: Algorithm/Go/977. Squares of a Sorted Array.go
/*
 * Created on Mon Jan 21 2019 22:52:12
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

 // two pointers, O(n) time, O(n) space
func abs(num int) int {
	if num <= 0 {
		return -1 * num
	} else {
		return num
	}
}

func sortedSquares(A []int) []int {
	idx := len(A)
	for i, num := range A {
		if num >= 0 {
			idx = i
			break
		}
	}
	result := []int{}
	left, right := idx - 1, idx
	for left>=0 && right<len(A) {
		if (abs(A[left]) > abs(A[right])) {
			result = append(result, A[right]*A[right])
			right++
		} else {
			result = append(result, A[left]*A[left])
			left--		
		}
	}
	for left>=0 {
		result = append(result, A[left]*A[left])
		left--		
	}
	for right<len(A) {
		result = append(result, A[right]*A[right])
		right++		
	}
	return result
}

File: Algorithm/C++/977. Squares of a Sorted Array.cc
/*
 * Created on Sun Jan 20 2019 18:22:36
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// O(n) time, O(n) space
#include <vector>
#include <cmath>

using std::vector;

class Solution {
 public:
  vector<int> sortedSquares(vector<int>& A) {
    int idx = A.size();
    for (int i = 0; i < A.size(); ++i) {
      if (A[i] >= 0) {
       idx = i;
       break;
      }
    }
    vector<int> result;
    int left = idx - 1, right = idx;
    while (left >= 0 && right < A.size()) {
      if (std::abs(A[left]) > std::abs(A[right])) {
        result.push_back(A[right] * A[right]);
        right++;
      } else {
        result.push_back(A[left] * A[left]);
        left--;        
      }
    }
    while (left >= 0) {
      result.push_back(A[left] * A[left]);
      left--;       
    }
    while (right < A.size()) {
      result.push_back(A[right] * A[right]);
      right++;      
    }
    return result;
  }
};